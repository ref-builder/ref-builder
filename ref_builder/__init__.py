"""Virtool CLI for managing references.

TODO: Switch back to Rich.
TODO: Allow exclusion checks at the Repo level.
TODO: Don't use the term "blacklist". Use the word "exclusions" only.
"""
