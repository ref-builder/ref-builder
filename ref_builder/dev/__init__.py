"""Development utilities for ref-builder."""
