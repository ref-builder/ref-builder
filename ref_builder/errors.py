from uuid import UUID


class LockConflictError(Exception):
    """Raised when a lock is attempted on a locked repository."""

    def __init__(self):
        super().__init__("Repository is already locked by another process.")


class LockRequiredError(Exception):
    """Raised when a transaction is attempted without a lock."""

    def __init__(self) -> None:
        super().__init__("Repository must be locked to use a transaction.")


class TransactionExistsError(Exception):
    """Raised when an error occurs during a transaction."""

    def __init__(self) -> None:
        super().__init__("An active transaction already exists.")


class TransactionRequiredError(Exception):
    """Raised when an error occurs during a transaction."""

    def __init__(self) -> None:
        super().__init__("Writing events requires a transaction and none was found.")


class PartialIDConflictError(Exception):
    """Raised when a partial ID used as a shortcode returns too many results."""

    def __init__(self) -> None:
        super().__init__("Found more than one result, may need longer partial.")


class InvalidInputError(Exception):
    """Raised when a command input does not match expected paramters."""

    def __init__(self, message: str) -> None:
        self.message = message

        super().__init__(message)


class OTUDeletedError(Exception):
    """Raised when an OTU is deleted."""

    def __init__(self, otu_id: UUID):
        super().__init__(f"OTU {otu_id} has been marked for deletion.")


class PlanConformationError(ValueError):
    """Raised when potential new sequences do not pass validation against
    the OTU plan.
    """
