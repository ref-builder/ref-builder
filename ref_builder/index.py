"""An index for improving repository performance."""

import binascii
import datetime
import sqlite3
from collections.abc import Iterator
from dataclasses import dataclass
from pathlib import Path
from uuid import UUID

from ref_builder.events.base import EventMetadata
from ref_builder.models.otu import OTUMinimal
from ref_builder.otu.builders.otu import OTUBuilder


@dataclass
class EventIndexItem:
    """The event IDs associated with an OTU."""

    event_ids: list[int]
    """A list of event IDs"""

    otu_id: UUID
    """The OTU ID this cache index is associated with."""


@dataclass
class Snapshot:
    """A snapshot of an OTU at a specific event."""

    at_event: int
    """The event at which the snapshot was taken."""

    otu: OTUBuilder
    """The OTU that was snapshotted."""


def adapt_uuid(original_uuid: UUID) -> str:
    """Automatically convert UUID to SQLite-compatible string."""
    return str(original_uuid)


def adapt_datetime(original_datetime: datetime.datetime) -> str:
    """Automatically convert datetime to SQLite-compatible string."""
    return original_datetime.isoformat()


def convert_uuid(string_uuid: bytes) -> UUID:
    """Automatically convert stringified UUID back to UUID type."""
    return UUID(string_uuid.decode())


def convert_datetime(string_datetime: bytes) -> datetime.datetime:
    """Automatically convert stringified datetime back to datetime type."""
    return datetime.datetime.fromisoformat(string_datetime.decode())


sqlite3.register_adapter(UUID, adapt_uuid)
sqlite3.register_adapter(datetime.datetime, adapt_datetime)

sqlite3.register_converter("uuid", convert_uuid)
sqlite3.register_converter("datetime", convert_datetime)


class Index:
    """An index for rapid access to repository data."""

    def __init__(self, path: Path) -> None:
        """Initialize the index."""
        self.path = path
        """The path to the index file."""

        self.path.parent.mkdir(exist_ok=True, parents=True)

        self.con = sqlite3.connect(
            path,
            isolation_level=None,
            detect_types=sqlite3.PARSE_DECLTYPES | sqlite3.PARSE_COLNAMES,
        )

        self.con.execute("PRAGMA journal_mode = WAL")
        self.con.execute("PRAGMA synchronous = NORMAL")

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS events (
                event_id INTEGER PRIMARY KEY,
                otu_id TEXT,
                timestamp TEXT
            );
            """,
        )

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS isolates (
                id TEXT PRIMARY KEY,
                name TEXT,
                otu_id TEXT
            );
            """,
        )

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS otus (
                id TEXT PRIMARY KEY,
                acronym TEXT,
                at_event INTEGER,
                name TEXT,
                otu JSONB,
                taxid INTEGER
            );
            """,
        )

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS sequences (
                id TEXT PRIMARY KEY,
                crc TEXT,
                otu_id TEXT,
                sequence TEXT
            );
            """,
        )

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS otu_updates (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                otu_id TEXT,
                timestamp_complete TEXT
            );
            """,
        )

        for table, column in [
            ("events", "otu_id"),
            ("isolates", "id"),
            ("otus", "id"),
            ("otus", "name"),
            ("otus", "taxid"),
            ("sequences", "otu_id"),
            ("sequences", "crc"),
            ("otu_updates", "otu_id"),
        ]:
            self.con.execute(
                f"""
                CREATE INDEX IF NOT EXISTS {table}_{column} ON {table} ({column});
                """,
            )

        self.con.commit()

    @property
    def otu_ids(self) -> set[UUID]:
        """A list of all OTUs tracked in the index."""
        return {
            row[0] for row in self.con.execute('SELECT id AS "id [uuid]" FROM otus')
        }

    def add_event_id(
        self,
        event_id: int,
        otu_id: UUID,
        timestamp: datetime.datetime,
    ) -> None:
        """Add a new event ID and associated it with a given OTU ID.

        This allows fast lookup of event IDs for a given OTU ID.

        :param event_id: the event ID to add
        :param otu_id: the OTU ID to add the event ID to
        :param timestamp: the timestamp of the event
        """
        self.con.execute(
            """
            INSERT INTO events (event_id, otu_id, timestamp)
            VALUES (?, ?, ?)
            """,
            (
                event_id,
                otu_id,
                timestamp,
            ),
        )

    def get_event_ids_by_otu_id(self, otu_id: UUID) -> EventIndexItem | None:
        """Get a list of event IDs for an OTU."""
        res = self.con.execute(
            """
            SELECT event_id
            FROM events
            WHERE otu_id = ?
            ORDER BY event_id
            """,
            (otu_id,),
        )

        event_ids = [row[0] for row in res]

        if event_ids:
            return EventIndexItem(
                event_ids,
                otu_id,
            )

        return None

    def get_first_timestamp_by_otu_id(self, otu_id: UUID) -> datetime.datetime | None:
        """Get the timestamp of the earliest event associated with this ID."""
        cursor = self.con.execute(
            """
            SELECT timestamp
            AS "timestamp [datetime]"
            FROM events
            WHERE otu_id = ?
            ORDER BY event_id
            """,
            (otu_id,),
        )

        if result := cursor.fetchone():
            return result[0]

        return None

    def get_latest_timestamp_by_otu_id(self, otu_id: UUID) -> datetime.datetime | None:
        """Get the timestamp of the latest event associated with this ID."""
        cursor = self.con.execute(
            """
            SELECT timestamp
            AS "timestamp [datetime]"
            FROM events
            WHERE otu_id = ?
            ORDER BY event_id DESC
            """,
            (otu_id,),
        )

        if result := cursor.fetchone():
            return result[0]

        return None

    def get_id_by_acronym(self, acronym: str) -> UUID | None:
        """Get an OTU ID by its acronym."""
        if not acronym:
            raise ValueError("Acronym must contain at least one character.")

        cursor = self.con.execute(
            'SELECT id AS "id [uuid]" FROM otus WHERE acronym = ?',
            (acronym,),
        )

        if result := cursor.fetchone():
            return result[0]

        return None

    def get_id_by_name(self, name: str) -> UUID | None:
        """Get an OTU ID by its name."""
        cursor = self.con.execute(
            'SELECT id AS "id [uuid]" FROM otus WHERE name = ?',
            (name,),
        )

        if result := cursor.fetchone():
            return result[0]

        return None

    def get_id_by_taxid(self, taxid: int) -> UUID | None:
        """Get an OTU ID by its taxonomy ID."""
        cursor = self.con.execute(
            'SELECT id AS "id [uuid]" FROM otus WHERE taxid = ?',
            (taxid,),
        )

        if result := cursor.fetchone():
            return result[0]

        return None

    def get_id_by_isolate_id(self, isolate_id: UUID) -> UUID | None:
        """Get an OTU ID from an isolate ID that belongs to it."""
        cursor = self.con.execute(
            'SELECT otu_id AS "otu_id [uuid]" FROM isolates WHERE id = ?',
            (isolate_id,),
        )

        if result := cursor.fetchone():
            return result[0]

        return None

    def delete_otu(self, otu_id: UUID) -> None:
        """Remove an OTU from the index.

        This happens rarely, so we just rewrite the whole index.

        :param otu_id: the ID of the OTU to remove.
        """
        self.con.execute(
            "DELETE FROM isolates WHERE otu_id = ?",
            (otu_id,),
        )

        self.con.execute(
            "DELETE FROM otus WHERE id = ?",
            (otu_id,),
        )

        self.con.execute(
            "DELETE FROM sequences WHERE otu_id = ?",
            (otu_id,),
        )

        self.con.commit()

    def iter_event_metadata(self) -> Iterator[EventMetadata]:
        """Iterate over event metadata."""
        rows = self.con.execute(
            "SELECT event_id, otu_id, timestamp FROM events ORDER BY event_id",
        ).fetchall()

        for row in rows:
            yield EventMetadata(
                id=row[0],
                otu_id=UUID(row[1]) if row[1] else None,
                timestamp=row[2],
            )

    def iter_minimal_otus(self) -> Iterator[OTUMinimal]:
        """Iterate over minimal representations of all OTUs in the index."""
        rows = self.con.execute(
            """
            SELECT acronym, id AS "id [uuid]", name, taxid
            FROM otus ORDER BY name
            """,
        ).fetchall()

        for row in rows:
            yield OTUMinimal(
                acronym=row[0],
                id=row[1],
                name=row[2],
                taxid=row[3],
            )

    def get_last_otu_update_timestamp(self, otu_id: UUID) -> datetime.datetime | None:
        """Get the timestamp of the last event in an update for an OTU."""
        cursor = self.con.execute(
            """
            SELECT timestamp_complete FROM otu_updates
            WHERE otu_id = ? ORDER BY id DESC
            """,
            (otu_id,),
        )

        if result := cursor.fetchone():
            return datetime.datetime.fromisoformat(result[0])

        return None

    def add_otu_update_history_entry(
        self, otu_id: UUID, timestamp: datetime.datetime
    ) -> int | None:
        """Write an entry into the OTU update history."""
        self.con.execute(
            """
            INSERT INTO
            otu_updates(otu_id, timestamp_complete)
            VALUES(?, ?)
            """,
            (
                otu_id,
                timestamp,
            ),
        )

        cursor = self.con.execute(
            "SELECT id FROM otu_updates WHERE otu_id = ? ORDER BY id DESC",
            (otu_id,),
        )

        if result := cursor.fetchone():
            return result[0]

        return None

    def load_snapshot(self, otu_id: UUID) -> Snapshot | None:
        """Load an OTU snapshot."""
        cursor = self.con.execute(
            "SELECT at_event, otu FROM otus WHERE id = ?",
            (otu_id,),
        )

        result = cursor.fetchone()

        if not result:
            return None

        at_event, otu_json = result

        otu = OTUBuilder.model_validate_json(otu_json)

        sequence_ids = [
            sequence.id for isolate in otu.isolates for sequence in isolate.sequences
        ]

        # Fetch all sequences in a single query
        placeholders = ",".join("?" for _ in sequence_ids)

        sequence_map = dict(
            self.con.execute(
                f"SELECT id, sequence FROM sequences WHERE id IN ({placeholders})",
                sequence_ids,
            ).fetchall(),
        )

        # Update the data structure and check for missing sequences
        missing_sequences = []

        for isolate in otu.isolates:
            for sequence in isolate.sequences:
                sequence_id = str(sequence.id)

                if sequence_id in sequence_map:
                    sequence.sequence = sequence_map[sequence_id]
                else:
                    missing_sequences.append(str(sequence_id))

        # Raise an error if any sequences are missing
        if missing_sequences:
            raise ValueError(f"Sequences not found: {', '.join(missing_sequences)}")

        return Snapshot(
            at_event=at_event,
            otu=OTUBuilder.model_validate(otu),
        )

    def upsert_otu(self, otu: OTUBuilder, at_event: int) -> None:
        """Update the index based on a list of OTUs.

        OTUs that aren't already in the index will be added, and those that are will be
        updated.

        Only sequences that have changed will be updated.
        """
        sequence_ids = {seq.id for isolate in otu.isolates for seq in isolate.sequences}

        placeholders = ",".join("?" for _ in sequence_ids)

        # Delete any sequences that are no longer in the OTU.
        self.con.execute(
            f"""
            DELETE FROM sequences WHERE otu_id = ? AND NOT id IN ({placeholders});
            """,
            (otu.id, *sequence_ids),
        )

        batch = []

        # Insert or update the isolates.
        for isolate in otu.isolates:
            self.con.execute(
                "INSERT OR REPLACE INTO isolates (id, name, otu_id) VALUES (?, ?, ?)",
                (
                    isolate.id,
                    str(isolate.name),
                    otu.id,
                ),
            )

            for sequence in isolate.sequences:
                crc = _calculate_crc32(sequence.sequence)
                batch.append(
                    (
                        sequence.id,
                        crc,
                        otu.id,
                        sequence.sequence,
                    ),
                )

        # Perform batch insert
        self.con.executemany(
            """
            INSERT OR REPLACE INTO sequences (id, crc, otu_id, sequence)
            VALUES (?, ?, ?, ?)
            """,
            batch,
        )

        # Update only if CRC has changed
        self.con.executemany(
            """
            UPDATE sequences
            SET crc = ?, otu_id = ?, sequence = ?
            WHERE id = ? AND crc != ?
            """,
            [(crc, otu_id, seq, seq_id, crc) for seq_id, crc, otu_id, seq in batch],
        )

        self.con.execute(
            """
            INSERT OR REPLACE INTO otus (id, acronym, at_event, name, otu, taxid)
            VALUES (?, ?, ?, ?, ?, ?)
            """,
            (
                otu.id,
                otu.acronym,
                at_event,
                otu.name,
                otu.model_dump_json(),
                otu.taxid,
            ),
        )

        self.con.commit()

    def prune(self, event_id: int) -> None:
        """Rollback the index to a previous event.

        * Deletes all events after the given event ID.
        * Deletes all OTU snapshots that have ``at_event`` later than ``event_id``.

        Removed snapshots will be automatically updated next time an OTU is fetched from
        the repository.

        :param event_id: the event ID to rollback to
        """
        self.con.execute(
            """
            DELETE FROM events WHERE event_id > ?
            """,
            (event_id,),
        )

        self.con.execute(
            """
            DELETE FROM otus WHERE at_event > ?
            """,
            (event_id,),
        )

        self.con.commit()


def _calculate_crc32(sequence: str) -> str:
    """Calculate the CRC32 checksum of a sequence.

    :param sequence: the sequence to calculate the checksum of.
    :return: the CRC32 checksum as a hexadecimal string.

    """
    crc = binascii.crc32(sequence.encode("utf-8"))

    # Convert CRC as a hexadecimal string.
    return hex(crc & 0xFFFFFFFF)[2:].zfill(8)
