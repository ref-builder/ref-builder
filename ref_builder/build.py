from pathlib import Path

import arrow
import orjson

from ref_builder.repo import Repo


def build_json(indent: bool, output_path: Path, path: Path, version: str) -> None:
    """Build a Virtool reference JSON file from a data directory.

    :param indent: whether to indent the JSON output
    :param output_path: The path to write the output JSON file to
    :param path: The path to a reference repository
    :param version: the version string to include in the reference.json file
    """
    repo = Repo(path)

    otus = []

    for otu in repo.iter_otus():
        isolates = []

        for isolate in otu.isolates:
            sequences = [
                {
                    "_id": sequence.legacy_id or sequence.id,
                    "accession": str(sequence.accession),
                    "definition": sequence.definition,
                    "host": "",
                    "segment": sequence.segment,
                    "sequence": sequence.sequence,
                }
                for sequence in isolate.sequences
            ]

            isolates.append(
                {
                    "id": isolate.legacy_id or isolate.id,
                    "default": isolate.id == otu.repr_isolate,
                    "sequences": sequences,
                    "source_name": isolate.name.value,
                    "source_type": isolate.name.type,
                },
            )

        isolates.sort(key=lambda x: x["id"])

        otus.append(
            {
                "_id": otu.legacy_id or otu.id,
                "abbreviation": otu.acronym,
                "isolates": isolates,
                "name": otu.name,
                "schema": otu.schema.model_dump(),
                "taxid": otu.taxid,
            },
        )

    otus.sort(key=lambda x: x["name"])

    with open(output_path, "wb") as f:
        f.write(
            orjson.dumps(
                {
                    "created_at": arrow.utcnow().isoformat(),
                    "data_type": repo.meta.data_type,
                    "name": version,
                    "organism": repo.meta.organism,
                    "otus": otus,
                },
                option=orjson.OPT_INDENT_2 if indent else 0,
            ),
        )
