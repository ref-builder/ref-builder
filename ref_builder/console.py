from collections.abc import Iterator

import rich.console
from rich.table import Table

from ref_builder.models import OTUMinimal
from ref_builder.plan import SegmentRule
from ref_builder.resources import RepoOTU


def _render_taxonomy_id_link(taxid: int) -> str:
    return f"[link=https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id={taxid}]{taxid}[/link]"


def _render_nucleotide_link(accession: str) -> str:
    return f"[link=https://www.ncbi.nlm.nih.gov/nuccore/{accession}]{accession}[/link]"


def print_otu(otu: RepoOTU) -> None:
    """Print the details for an OTU to the console.

    :param otu: The OTU to print.

    """
    console.print(f"[bold][underline]{otu.name}[/bold][/underline]")
    console.line()

    table = Table(
        box=None,
        show_header=False,
    )

    table.add_row("[bold]ACRONYM[/bold]", otu.acronym)
    table.add_row("[bold]ID[/bold]", str(otu.id))

    if otu.legacy_id:
        table.add_row("[bold]LEGACY ID[/bold]", otu.legacy_id)

    table.add_row("[bold]TAXID[/bold]", _render_taxonomy_id_link(otu.taxid))

    max_accession_length = max(
        len(str(sequence.accession))
        for isolate in otu.isolates
        for sequence in isolate.sequences
    )

    max_segment_name_length = max(
        len(str(segment.name)) for segment in otu.plan.segments
    )

    console.print(table)

    console.line()
    console.print("[bold]PLAN[/bold]")
    console.line()

    plan_table = Table(box=None)

    plan_table.add_column("NAME")
    plan_table.add_column("REQUIRED")
    plan_table.add_column("LENGTH")
    plan_table.add_column("TOLERANCE")
    plan_table.add_column("ID")

    if not otu.plan.monopartite:
        for segment in otu.plan.segments:
            plan_table.add_row(
                str(segment.name),
                "[red]Yes[/red]"
                if segment.required == SegmentRule.REQUIRED
                else "[grey]No[/grey]",
                str(segment.length),
                str(segment.length_tolerance),
                str(segment.id),
            )
    else:
        monopartite_segment = otu.plan.segments[0]

        segment_name = (
            monopartite_segment.name
            if monopartite_segment.name is not None
            else "Unnamed"
        )

        plan_table.add_row(
            segment_name,
            "[red]Yes[/red]",
            str(monopartite_segment.length),
            str(monopartite_segment.length_tolerance),
        )

    console.print(plan_table)

    console.line()
    console.print("[bold]ISOLATES[/bold]")

    for isolate in otu.isolates:
        console.line()
        console.print(str(isolate.name)) if isolate.name is not None else console.print(
            "[UNNAMED]",
        )
        console.line()

        isolate_table = Table(
            box=None,
        )

        isolate_table.add_column("ACCESSION", width=max_accession_length)
        isolate_table.add_column("LENGTH")
        isolate_table.add_column("SEGMENT", min_width=max_segment_name_length)
        isolate_table.add_column("DEFINITION")

        for sequence in sorted(isolate.sequences, key=lambda s: s.accession):
            isolate_table.add_row(
                _render_nucleotide_link(str(sequence.accession)),
                str(len(sequence.sequence)),
                str(otu.plan.get_segment_by_id(sequence.segment).name),
                sequence.definition,
            )

        console.print(isolate_table)


def print_otu_list(otus: Iterator[OTUMinimal]) -> None:
    """Print a list of OTUs to the console.

    :param otus: The OTUs to print.

    """
    table = Table(
        box=None,
    )

    table.add_column("NAME")
    table.add_column("ACRONYM")
    table.add_column("TAXID")
    table.add_column("ID")

    added_rows = False

    for otu in otus:
        table.add_row(
            otu.name,
            otu.acronym,
            str(otu.taxid),
            str(otu.id),
        )

        added_rows = True

    if added_rows:
        console.print(table)
    else:
        console.print("No OTUs found")


console = rich.console.Console()
