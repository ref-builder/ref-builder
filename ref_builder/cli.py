"""Command-line interface for reference builder."""

import glob
import sys
from pathlib import Path
from uuid import UUID

import click

from ref_builder.build import build_json
from ref_builder.console import print_otu, print_otu_list
from ref_builder.legacy.convert import convert_legacy_repo
from ref_builder.legacy.utils import iter_legacy_otus
from ref_builder.legacy.validate import validate_legacy_repo
from ref_builder.logs import configure_logger
from ref_builder.ncbi.client import NCBIClient
from ref_builder.options import debug_option, ignore_cache_option, path_option
from ref_builder.otu.create import create_otu
from ref_builder.otu.update import add_isolate, auto_update_otu
from ref_builder.repo import Repo
from ref_builder.utils import DataType, format_json


@click.group()
def entry() -> None:
    """Build and maintain reference sets of pathogen genome sequences."""


@entry.command()
@click.option(
    "--data-type",
    help="the type of data the reference contains (eg. genome)",
    required=True,
    type=click.Choice(DataType),
)
@click.option(
    "--name",
    help="the type of data the reference contains (eg. genome)",
    required=True,
    type=str,
)
@click.option(
    "--organism",
    default="",
    help="the organism the reference is for (eg. virus)",
    type=str,
)
@click.option(
    "--path",
    default=".",
    help="the path to initialize the repository at",
    type=click.Path(path_type=Path),
)
def init(data_type: DataType, name: str, organism: str, path: Path) -> None:
    """Create a new event-sourced repo."""
    Repo.new(data_type, name, path, organism)


@entry.group()
def otu() -> None:
    """Manage OTUs."""


@otu.command(name="create")
@click.argument("TAXID", type=int)
@click.argument(
    "accessions_",
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.option("--acronym", type=str, default="")
@click.option("--autofill/--no-fill", default=False)
@ignore_cache_option
@debug_option
@path_option
def otu_create(
    debug: bool,
    ignore_cache: bool,
    path: Path,
    taxid: int,
    accessions_: list[str],
    autofill: bool,
    acronym: str,
) -> None:
    """Create a new OTU for the given Taxonomy ID and accessions."""
    configure_logger(debug)

    repo = Repo(path)

    try:
        new_otu = create_otu(
            repo,
            taxid,
            accessions_,
            acronym=acronym,
            ignore_cache=ignore_cache,
        )
    except ValueError as e:
        click.echo(e, err=True)
        sys.exit(1)

    if autofill:
        auto_update_otu(repo, new_otu, ignore_cache=ignore_cache)


@otu.command(name="get")
@click.argument("IDENTIFIER", type=str)
@path_option
def otu_get(identifier: str, path: Path) -> int:
    """Get an OTU by its Taxonomy ID."""
    try:
        identifier = int(identifier)
        otu_ = Repo(path).get_otu_by_taxid(identifier)
    except ValueError:
        otu_ = Repo(path).get_otu(UUID(identifier))

    if otu_ is None:
        click.echo("No OTU found.", err=True)
        return 1

    print_otu(otu_)

    return 0


@otu.command(name="list")
@path_option
def otu_list(path: Path) -> None:
    """List all OTUs in the repository."""
    print_otu_list(Repo(path).iter_otus())


@otu.group(invoke_without_command=True)
@click.argument("TAXID", type=int)
@path_option
@click.pass_context
def update(ctx, path: Path, taxid: int):
    """Update the specified OTU with new data."""
    ctx.ensure_object(dict)

    ctx.obj['TAXID'] = taxid

    repo = Repo(path)

    ctx.obj['REPO'] = repo

    otu_id = repo.get_otu_id_by_taxid(taxid)
    if otu_id is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    if ctx.invoked_subcommand is None:
        click.echo(ctx.get_help())


@update.command(name="automatic")
@debug_option
@ignore_cache_option
@click.pass_context
def otu_autoupdate(ctx, debug: bool, ignore_cache: bool) -> None:
    """Automatically update an OTU with the latest data from NCBI."""
    configure_logger(debug)

    taxid = ctx.obj['TAXID']

    repo = ctx.obj['REPO']

    otu_ = repo.get_otu_by_taxid(taxid)

    if otu_ is None:
        click.echo(f"OTU not found for Taxonomy ID {taxid}.", err=True)
        click.echo(f'Run "virtool otu create {taxid} --autofill" instead.')
        sys.exit(1)

    auto_update_otu(repo, otu_, ignore_cache=ignore_cache)


@update.command(name="isolate")
@click.argument(
    "accessions_",
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@ignore_cache_option
@debug_option
@click.pass_context
def isolate_create(
    ctx,
    debug: bool,
    ignore_cache: bool,
    accessions_: list[str],
) -> None:
    """Create a new isolate using the given accessions."""
    configure_logger(debug)

    repo = ctx.obj['REPO']

    taxid = ctx.obj['TAXID']

    otu_ = repo.get_otu_by_taxid(taxid)
    if otu_ is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    try:
        add_isolate(
            repo,
            otu_,
            accessions_,
            ignore_cache=ignore_cache,
        )
    except ValueError as e:
        click.echo(e, err=True)
        sys.exit(1)


@entry.group()
def legacy() -> None:
    """Validate and convert legacy references."""


@legacy.command()
@click.option(
    "--name",
    help="the name to set for the new repository",
    required=True,
    type=str,
)
@click.option(
    "--path",
    help="the path for the legacy reference repository",
    required=True,
    type=click.Path(exists=True, file_okay=False, path_type=Path),
)
@click.option(
    "--target-path",
    help="the path for the new converted repository",
    required=True,
    type=click.Path(exists=False, file_okay=False, path_type=Path),
)
def convert(name: str, path: Path, target_path: Path) -> None:
    """Convert a legacy reference repository to a new event-sourced repository."""
    convert_legacy_repo(name, path, target_path)


@legacy.command()
@click.option(
    "--path",
    help="the name to the legacy reference repository",
    required=True,
    type=click.Path(exists=True, file_okay=False, path_type=Path),
)
def precache(path: Path) -> None:
    """Pre-cache all accessions in a legacy reference repository."""
    ncbi = NCBIClient(False)

    buffer = []

    for otu_ in iter_legacy_otus(path / "src"):
        for isolate in otu_["isolates"]:
            for sequence in isolate["sequences"]:
                buffer.append(sequence["accession"])

                if len(buffer) > 450:
                    ncbi.fetch_genbank_records(buffer)
                    buffer = []

    if buffer:
        ncbi.fetch_genbank_records(buffer)


@legacy.command(name="format")
@click.option(
    "--path",
    help="the name to the legacy reference repository",
    required=True,
    type=click.Path(exists=True, file_okay=False, path_type=Path),
)
def reformat(path: Path) -> None:
    """Format a legacy reference repository.

    Re-formats every JSON file in a legacy reference repository to have a consistent
    format.
    """
    src_path = path / "src"

    for json_path in glob.glob(str(src_path / "**/*.json"), recursive=True):
        format_json(Path(json_path))


@legacy.command()
@debug_option
@click.option(
    "--fix",
    is_flag=True,
    help="attempt to fix errors in-place",
)
@click.option(
    "--limit",
    default=0,
    help="exit if this many otus have errors",
)
@click.option(
    "--no-ok",
    is_flag=True,
    help="don't print anything if an otu is valid",
)
@click.option(
    "--path",
    required=True,
    type=click.Path(exists=True, file_okay=False, path_type=Path),
    help="the path to a legacy reference directory",
)
def validate(debug: bool, fix: bool, limit: int, no_ok: bool, path: Path) -> None:
    """Validate a legacy reference repository."""
    configure_logger(debug)

    validate_legacy_repo(fix, limit, no_ok, path)


@entry.command()
@click.option(
    "-i",
    "--indent",
    is_flag=True,
    help="auto-indent the output JSON file",
)
@click.option(
    "-V",
    "--version",
    default="",
    type=str,
    help="a version string to include in the reference.json file",
)
@click.option(
    "-o",
    "--output-path",
    required=True,
    type=click.Path(exists=False, file_okay=True, path_type=Path),
    help="the path to write the reference.json file to",
)
@click.option(
    "-p",
    "--path",
    required=True,
    type=click.Path(exists=True, file_okay=False, path_type=Path),
    help="the path to the reference repository",
)
def build(output_path: Path, path: Path, indent: bool, version: str) -> None:
    """Build a Virtool reference.json file from a reference repository."""
    build_json(indent, output_path, path, version)
