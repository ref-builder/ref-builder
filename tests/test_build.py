import json
from pathlib import Path

import arrow
from syrupy import SnapshotAssertion
from syrupy.filters import props

from ref_builder.build import build_json


def test_ok(
    scratch_path: Path,
    snapshot: SnapshotAssertion,
    tmp_path: Path,
):
    """Test that the command exits with a 0 exit code."""
    output_path = tmp_path / "reference.json"

    build_json(False, output_path, scratch_path, "2.1.0")

    with open(output_path) as f:
        built_json = json.load(f)

    assert built_json == snapshot(exclude=props("created_at"))
    assert built_json["name"] == "2.1.0"
    assert (arrow.utcnow() - arrow.get(built_json["created_at"])).seconds == 0


def test_indent(scratch_path: Path, tmp_path: Path):
    """Test that the indent in the reference.json file is properly set."""
    output_path = tmp_path / "reference.json"
    output_indented_path = tmp_path / "reference_indent.json"

    build_json(False, output_path, scratch_path, "2.1.0")
    build_json(True, output_indented_path, scratch_path, "2.1.0")

    assert {**json.load(output_path.open("rb")), "created_at": ""} == {
        **json.load(output_indented_path.open("rb")),
        "created_at": "",
    }
