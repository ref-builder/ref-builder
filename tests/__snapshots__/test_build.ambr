# serializer version: 1
# name: test_ok
  dict({
    'data_type': 'genome',
    'name': '2.1.0',
    'organism': 'viruses',
  })
# ---
